UIforETW/UIforETWDlg.cpp

```diff
@@ -40,10 +41,821 @@ const int kRecordTraceHotKey = 1234;
+	//_wgetenv_s can be a bit weird.
+	//The parameter is documented as (returning):
+	//The buffer size that's required, or 0 if the variable is not found.
+	const errno_t getEnvironmentVariableLengthResult = _wgetenv_s(&returnValue_temp, NULL, 0, env);
+	const rsize_t returnValue = returnValue_temp;
+	if (getEnvironmentVariableLengthResult != 0)
+	{
+		ATLASSERT(getEnvironmentVariableLengthResult != ERANGE);
+		ATLASSERT(getEnvironmentVariableLengthResult != EINVAL);
+		outputPrintf(L"Failed to get value (expected a directory)"
+					 L"of environment variable `%s`!!\n", env);
+		return L"";
+	}
+	if (returnValue == 0)
+	{
+#ifdef DEBUG
+		outputPrintf(L"Environment variable `%s` not found - "
+					 L"(will use default directory).\n", env);
+#endif
+		return L"";
+	}
+
+	// Get a directory (from an environment variable, if set) and make sure it exists.
+	//std::wstring result = default;
+	const rsize_t bufferSize = 512u;
+	if ((returnValue + 1) > bufferSize)
+	{
+		rsize_t dynReturnValue = 0;
+		std::unique_ptr<wchar_t[ ]> dynamicBuffer =
+			std::make_unique<wchar_t[ ]>(returnValue + 1);
+		const errno_t getEnvResult =
+			_wgetenv_s(&dynReturnValue, dynamicBuffer.get(),
+					   (returnValue + 1), env);
+		ATLASSERT((wcslen(dynamicBuffer.get()) + 1) == dynReturnValue);
+		if (getEnvResult != 0)
+		{
+			ATLASSERT(getEnvResult != ERANGE);
+			ATLASSERT(getEnvResult != EINVAL);
+			outputPrintf(L"Failed to get value (expected a directory)"
+						 L"of environment variable `%s`!!\n", env);
+			return L"";
+		}
+		return dynamicBuffer.get();
+	}
+
+	rsize_t stackbufferReturnValue = 0;
+	wchar_t environmentVariableBuffer[bufferSize] = { 0 };
+	const errno_t getEnvResult = _wgetenv_s(&stackbufferReturnValue, environmentVariableBuffer, env);
+	ATLASSERT((wcslen(environmentVariableBuffer ) + 1) == stackbufferReturnValue);
+	if (getEnvResult != 0)
+	{
+		ATLASSERT(getEnvResult != ERANGE);
+		ATLASSERT(getEnvResult != EINVAL);
+		outputPrintf(L"Failed to get value (expected a directory)"
+					 L"of environment variable `%s`!!\n", env );
+		return L"";
+	}
+	return environmentVariableBuffer;
+}
+
+std::wstring getTraceDirFromEnvironmentVariable( _In_z_ PCWSTR env, const std::wstring& default )
+{
+	std::wstring rawDirectory = getRawDirectoryFromEnvironmentVariable(env);
+	std::wstring result = default;
+
+	if (!rawDirectory.empty())
+		result = rawDirectory;
+
+	// Make sure the name ends with a backslash.
+	if ((!result.empty()) && (result[ result.size( ) - 1 ] != '\\'))
+		result += '\\';
+	return result;
+}
+
+
+bool enhancedCreateDirectory( _In_ std::wstring pathName )
+{
+	//(first parameter):
+	//    There is a default string size limit for paths of 248 characters.
+	//    To extend this limit to 32,767 wide characters,
+	//        call the Unicode version of the function
+	//        and prepend "\\?\" to the path.
+	//                    ^ ==> L"\\\\?\\"
+
+	//TODO: what if network path?
+	//should we check if just starts with L"\\\\"?

```

Comment:
What is this doing? Why would the passed in path not be fine? What situation are you handling?